external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java

```diff
@@ -0,0 +1,654 @@
+    LOG.debug("Emitting - {}", id);
+    this.collector.emit(tuple, id);
+    inflight.put(id, tuple);
+  }
+
+  public void open(Map conf, TopologyContext context,  SpoutOutputCollector collector) {
+    this.conf = conf;
+    final String FILE_SYSTEM = "filesystem";
+    LOG.info("Opening");
+    this.collector = collector;
+    this.hdfsConfig = new Configuration();
+    this.tupleCounter = 0;
+
+    for( Object k : conf.keySet() ) {
+      String key = k.toString();
+      if( ! FILE_SYSTEM.equalsIgnoreCase( key ) ) { // to support unit test only
+        String val = conf.get(key).toString();
+        LOG.info("Config setting : " + key + " = " + val);
+        this.hdfsConfig.set(key, val);
+      }
+      else
+        this.hdfs = (FileSystem) conf.get(key);
+
+      if(key.equalsIgnoreCase(Configs.READER_TYPE)) {
+        readerType = conf.get(key).toString();
+        checkValidReader(readerType);
+      }
+    }
+
+    // - Hdfs configs
+    this.hdfsConfig = new Configuration();
+    Map<String, Object> map = (Map<String, Object>)conf.get(this.configKey);
+    if(map != null){
+      for(String key : map.keySet()){
+        this.hdfsConfig.set(key, String.valueOf(map.get(key)));
+      }
+    }
+
+    try {
+      HdfsSecurityUtil.login(conf, hdfsConfig);
+    } catch (IOException e) {
+      LOG.error("Failed to open " + sourceDirPath);
+      throw new RuntimeException(e);
+    }
+
+    // -- source dir config
+    if ( !conf.containsKey(Configs.SOURCE_DIR) ) {
+      LOG.error(Configs.SOURCE_DIR + " setting is required");
+      throw new RuntimeException(Configs.SOURCE_DIR + " setting is required");
+    }
+    this.sourceDirPath = new Path( conf.get(Configs.SOURCE_DIR).toString() );
+
+    // -- archive dir config
+    if ( !conf.containsKey(Configs.ARCHIVE_DIR) ) {
+      LOG.error(Configs.ARCHIVE_DIR + " setting is required");
+      throw new RuntimeException(Configs.ARCHIVE_DIR + " setting is required");
+    }
+    this.archiveDirPath = new Path( conf.get(Configs.ARCHIVE_DIR).toString() );
+
+    try {
+      if(hdfs.exists(archiveDirPath)) {
+        if(! hdfs.isDirectory(archiveDirPath) ) {
+          LOG.error("Archive directory is a file. " + archiveDirPath);
+          throw new RuntimeException("Archive directory is a file. " + archiveDirPath);
+        }
+      } else if(! hdfs.mkdirs(archiveDirPath) ) {
+        LOG.error("Unable to create archive directory. " + archiveDirPath);
+        throw new RuntimeException("Unable to create archive directory " + archiveDirPath);
+      }
+    } catch (IOException e) {
+      LOG.error("Unable to create archive dir ", e);
+      throw new RuntimeException("Unable to create archive directory ", e);
+    }
+
+    // -- bad files dir config
+    if ( !conf.containsKey(Configs.BAD_DIR) ) {
+      LOG.error(Configs.BAD_DIR + " setting is required");
+      throw new RuntimeException(Configs.BAD_DIR + " setting is required");
+    }
```

Comment:
Configs.BAD_DIR, Configs.ARCHIVE_DIR and Configs.SOURCE_DIR validation cheks are same. You may want to extract a method for these common validations.